draft-ion.adoc

```diff
@@ -208,79 +200,176 @@ represented as follows:
+
+If a member name is not equal to the case-sensitive octet sequence `meta`, an Ion parser MUST NOT identify a member as a _Meta Object_.
+
+The member names within an identified _Meta Object_ MUST be unique; Ion parsers MUST either
+reject _Meta Object_ members with duplicate member names or use a JSON
+parser that returns only the lexically last duplicate member, as specified
 in {uri-ecmaScript-5-1-15-12}[Section 15.12 (The JSON Object)] of {uri-ecmaScript-5-1}[ECMAScript 5.1].
 
-The set of fields that a _Meta Object_ must contain to be considered valid
+The set of members that a _Meta Object_ must contain to be considered valid
 is resource-dependent.  In the absense of any resource-specific requirements,
-any field not understood by implementations MUST be ignored.
+any member not understood by implementations MUST be ignored.
+
+[[links]]
+== Links
+
+An Ion Link is an Ion Object where its `meta` member contains an {uri-IRIs}[IRI] member named `href`. An Ion Link is therefore JSON serialization of a {uri-rfc5988}[web link] to another resource.
+
+Ion parsers MUST identify any JSON structure matching this definition as an Ion Link.  Anything that does not match this defined structure MUST NOT be considered an Ion Link.
+
+A non-normative example of a minimal Ion link:
+
+.Example simplest Ion link:
+[source,json]
+----
+{ "meta": { "href": "https://api.example.com/corporations/acme" } }
+----
+
+=== Link Relation Types
+
+All Ion links have one or more <<links, link relation types>>.  A link's total set of link relation types is the set union of the link's implicit link relation type and the link's explicit link relation types.
+
+==== Implicit Link Relation Type
+
+All Ion links have exactly one implicit link relation type.  The implicit link relation type can be one of three possible values depending on where the link is located in the Ion content:
+
+1. Root Object
+2. Named Link
+3. Collection Resource `items` array element
+
+===== Root Object Implicit Link Relation Type
+
+If the Root Object is also found to be an Ion Link, it has an implicit link relation type of `self`.
+
+A non-normative example of a Root Object that is also an Ion Link:
+
+[source,json]
+----
+{
+  "meta": { "href": "https://api.example.com/users/1" },
+  "name": "Joe"
+}
+----
 
-=== Registered Metadata Field Names
+===== Named Link Implicit Link Relation Type
 
-The following field names are registered in the IANA Ion Metadata field registry
-defined in [TBD].  None of the fields defined below are intended to be mandatory
+If a JSON Object member value is an Ion Link, the Ion Link is implicitly assigned a link relation type that equals the member name.
+
+A non-normative example:
+
+[source,json]
+----
+{
+  "employer": { "meta": { "href": "https://api.example.com/corporations/acme" } }
+}
+----
+
+The above example shows an Ion Link member named `employer`.  The octet sequence `employer` is therefore assigned as the link's implicit link relation type.
```

Comment:
> I think the lines are blurring here and there is an aspect of education on hypermedia design principles feeling like they are being baked into the spec. But I don't think they belong there.

This is great information and really helpful for me to understand your arguments.  I think the the distinction between media type semantics vs domain semantics and this exactly reflects the LoggIt example I gave above, and I agree totally with the distinction.

When I thought of Ion, as I do (mostly) now, I don't want any specific domain semantics built in to the media type, but I *did*/do want to ensure that it would facilitate best practices of REST design.  This is what people have the most problems with these days, and is (I believe) where people are looking for strong support.

In other words, I don't think hand-holding or being opinionated is such a bad thing sometimes.  For example, let's say that Ion *did* mandate that Ion content `MUST NOT` contain any JSON array other than the `items` array in what I call a Collection Resource.

Is this absolutely mandatory for a hypermedia media type? No, because hypermedia is all about eliminating out-of-band information and ensuring links/transitions.  As long as you do that (reasonably well), you have a hypermedia type.

But by having this in Ion, you ensure, to a stronger degree, that people won't break their domain model, because Ion is more resilient to data type changes over time than without this mandate.  It makes an even stronger guarantee of backwards-compatible evolvability than without it, and that's great for REST APIs, especially as it appears to me that many people building REST APIs don't do this well at all.

So I guess my question is - and I know we'll dig in to this when we meet for lunch, so feel free to not answer now - what would be the harm in adding in these additional things that help make domain semantics stronger?  There is no particular domain baked in, just common things that are better for all domains.

Food for thought!